### PR TITLE
Babel 7

### PR DESCRIPTION
**Pull request by Hypnosphi**
Probably needs delaying: https://github.com/storybooks/storybook/pull/3746#issuecomment-403150140

DISCLAIMER: this is about moving Storybook build process itself to Babel 7, and making it the default. You should be able to use Babel 7 in your app with current stable version of Storybook, you just need to install [babel-bridge](https://github.com/babel/babel-bridge)

Issue: #3335 

This doesn't yet fix all the issues linked from there but it performs the necessary breaking change so that rest of the work could be done in a minor release

### CODE REVIEW

**Comment by vicapow** [comment]
Woo!! Can't wait for this to land. Anything we can do to help?

**Comment by Hypnosphi** [comment]
Thanks @vicapow, I hope to be able to finish this in upcoming days

**Comment by augbog** [comment]
Thanks for all your hard work on this -- looking forward to this :)

**Comment by ndelangen** [comment]
@vicapow When this lands in the first alpha, we're going to need a brave soul to test it 😎

**Comment by sakulstra** [comment]
@ndelangen we're definitely going to :)
storybook is the last tool in our chain hopping on the babel 7 beta train :D so this will have quite a positive affect on our package.lock

**Comment by vicapow** [comment]
Happy to volunteer!

**Comment by Hypnosphi** [comment]
We have to use `babel-preset-minify@canary` for now, see https://github.com/babel/minify/issues/850

**Comment by Hypnosphi** [comment]
@vicapow @augbog I've added a disclamer on top

**Comment by Hypnosphi** [comment]
Things left to do here:

* [ ] RN
* [ ] Jest

**Comment by Hypnosphi** [comment]
Looks like https://github.com/facebook/jest/issues/6573 is a blocker here

File: lib/cli/test/snapshots/vue/package.json
Comment (shilman) [inline lib/cli/test/snapshots/vue/package.json]: @Hypnosphi this looks suspicious?

File: lib/components/src/addon_panel/__snapshots__/index.stories.storyshot
Comment (shilman) [inline lib/components/src/addon_panel/__snapshots__/index.stories.storyshot]: @Hypnosphi what's up with all these extra CSS classes? Where do they come from? 🤔

File: lib/cli/test/snapshots/vue/package.json
Comment (Hypnosphi) [reply lib/cli/test/snapshots/vue/package.json]: I add it inside app, to be consistent with `@storybook/react`:

https://github.com/storybooks/storybook/pull/3746/files#diff-48d46ebd591f6220a5142b66a694395b

File: lib/components/src/addon_panel/__snapshots__/index.stories.storyshot
Comment (Hypnosphi) [reply lib/components/src/addon_panel/__snapshots__/index.stories.storyshot]: No idea, something from emotion probably

**Comment by Hypnosphi** [comment]
Looks like we can't migrate RN right now. It only supports Babel 7 since yesterday, and tools like CRNA and Expo haven't adapted yet. Let's do this in a separate PR

File: lib/components/src/addon_panel/__snapshots__/index.stories.storyshot
Comment (Hypnosphi) [reply lib/components/src/addon_panel/__snapshots__/index.stories.storyshot]: This PR updates `babel-plugin-emotion` among other things, so that may be it

**Comment by Hypnosphi** [comment]
I think I'll add a peer dependency on `babel-loader` for easier backwards compatibility

**Comment by Hypnosphi** [comment]
Babel deprecates stage presets, so we need to pick the particular plugins we want to add, see https://github.com/babel/babel/releases/tag/v7.0.0-beta.52

Looking at the number of breaking changes in each beta, I tend to think that we need to hold off this PR until stable Babel 7 is released. There are no estimations for that, so Storybook v4 should probably be released without this PR

CC @shilman @ndelangen

**Comment by NameFILIP** [comment]
😬😬😬

**Comment by kirill-konshin** [comment]
Just make sure Storybook can work without quirks on a project based on Babel 7. I suggest to somehow make it easily pluggable. Currently we can reconfigure Webpack, but that's quite awkward.

**Comment by Hypnosphi** [comment]
Storybook should work as is with babel-bridge, given that you have a `.babelrc` file

**Comment by sneko** [comment]
@Hypnosphi before I had this error:
```
info @storybook/vue v3.4.8
info
=> Loading custom .babelrc
=> Loading custom addons config.
=> Using default webpack setup based on "vue-cli".
info Building storybook ...
ERR! Failed to build the storybook
ERR! ./.storybook/config.js
ERR! Module build failed: TypeError: Cannot read property 'bindings' of null
ERR!     at Scope.moveBindingTo (E:\XXXXXXXXXXXXXXXXXX\node_modules\@babel\traverse\lib\scope\index.js:978:13)
ERR!     at BlockScoping.updateScopeInfo (E:\XXXXXXXXXXXXXXXXXX\node_modules\babel-plugin-transform-es2015-block-scoping\lib\index.js:364:17)
ERR!     at BlockScoping.run (E:\XXXXXXXXXXXXXXXXXX\node_modules\babel-plugin-transform-es2015-block-scoping\lib\index.js:330:12)
ERR!     at PluginPass.BlockStatementSwitchStatementProgram (E:\XXXXXXXXXXXXXXXXXX\node_modules\babel-plugin-transform-es2015-block-scoping\lib\index.js:70:24)
ERR!     at newFn (E:\XXXXXXXXXXXXXXXXXX\node_modules\@babel\traverse\lib\visitors.js:237:21)
ERR!     at NodePath._call (E:\XXXXXXXXXXXXXXXXXX\node_modules\@babel\traverse\lib\path\context.js:65:20)
ERR!     at NodePath.call (E:\XXXXXXXXXXXXXXXXXX\node_modules\@babel\traverse\lib\path\context.js:40:17)
ERR!     at NodePath.visit (E:\XXXXXXXXXXXXXXXXXX\node_modules\@babel\traverse\lib\path\context.js:100:12)
ERR!     at TraversalContext.visitQueue (E:\XXXXXXXXXXXXXXXXXX\node_modules\@babel\traverse\lib\context.js:142:16)
ERR!     at TraversalContext.visitSingle (E:\XXXXXXXXXXXXXXXXXX\node_modules\@babel\traverse\lib\context.js:102:19)
ERR!     at TraversalContext.visit (E:\XXXXXXXXXXXXXXXXXX\node_modules\@babel\traverse\lib\context.js:182:19)
ERR!     at Function.traverse.node (E:\XXXXXXXXXXXXXXXXXX\node_modules\@babel\traverse\lib\index.js:106:17)
ERR!     at traverse (E:\XXXXXXXXXXXXXXXXXX\node_modules\@babel\traverse\lib\index.js:76:12)
ERR!     at transformFile (E:\XXXXXXXXXXXXXXXXXX\node_modules\@babel\core\lib\transformation\index.js:116:29)
ERR!     at runSync (E:\XXXXXXXXXXXXXXXXXX\node_modules\@babel\core\lib\transformation\index.js:45:3)
ERR!     at transformSync (E:\XXXXXXXXXXXXXXXXXX\node_modules\@babel\core\lib\transform-sync.js:17:38)
ERR!  @ multi ./node_modules/@storybook/vue/dist/server/config/polyfills.js ./node_modules/@storybook/vue/dist/server/config/globals.js ./.storybook/config.js
```

After looking on forums I understand that there was an issue of babel versions. Now I'm trying to set the "bridge" by following their README.md:
https://github.com/babel/babel-bridge

But I get know this kind of error:
```
info @storybook/vue v3.4.8
info
=> Loading custom .babelrc
=> Loading custom addons config.
=> Using default webpack setup based on "vue-cli".
info Building storybook ...
ERR! Failed to build the storybook
ERR! ./node_modules/@storybook/vue/dist/client/preview/NoPreview.vue?vue&type=style&index=0&lang=css
ERR! Module parse failed: Unexpected token (21:0)
ERR! You may need an appropriate loader to handle this file type.
ERR! |
ERR! |
ERR! | .nopreview_wrapper {
ERR! |   position: fixed;
ERR! |   display: flex;
ERR!  @ ./node_modules/@storybook/vue/dist/client/preview/NoPreview.vue 4:0-68
ERR!  @ ./node_modules/@storybook/vue/dist/client/preview/render.js
ERR!  @ ./node_modules/@storybook/vue/dist/client/preview/index.js
ERR!  @ ./node_modules/@storybook/vue/dist/client/index.js
ERR!  @ ./.storybook/config.js
ERR!  @ multi ./node_modules/@storybook/vue/dist/server/config/polyfills.js ./node_modules/@storybook/vue/dist/server/config/globals.js ./.storybook/config.js
```

Is there potentially something I forgot to do?

Thanks

**Comment by ndelangen** [comment]
I'm trying to get this back into action, merging in master and testing..

**Review by ndelangen — APPROVED** [review]

**Comment by ndelangen** [comment]
@Hypnosphi I think it's green now. Do you have a list of remaining tasks on this or after this?

**Comment by pksunkara** [comment]
Guys, there's some ambiguity regarding this. Can someone confirm whether this will be part of v4 or not?

**Comment by ndelangen** [comment]
@pksunkara we're discussing what to do. I got recommended by the babel team to hold off releasing this as latest until they release a release candidate.

We might do duo releases until then, @shilman seemed to be up for doing that. It seems to be quite in demand 👍.

But obviously the reason we're careful here is because this could break a lot of people's setup and make it hard for them to upgrade. Considering the babel team has (AFAIK) 2 more breaking changes planned in the beta, we don't want to burden our users with a lot of migrations.

@Hypnosphi proposed to wait until create-react-app migrates too.

**Comment by pksunkara** [comment]
> We might do duo releases until then

Something like `storybook@babel7`?

**Review by danielduan — APPROVED** [review]

**Comment by danielduan** [comment]
I'm gonna start on https://github.com/storybooks/babel-plugin-react-docgen/pull/38 then

**Comment by ndelangen** [comment]
> Something like storybook@babel7?
Exactly!

☝️? @shilman @Hypnosphi

**Comment by Hypnosphi** [comment]
`^4.0.0-babel7` will match `4.0.0` which is probably not what we want.

And we can't move from 6 to 7 without a breaking change, so it probably should just be `@storybook/react@5.0.0-alpha`

**Comment by chrisnojima** [comment]
@ndelangen just FYI: https://github.com/babel/babel/releases/tag/v7.0.0-rc.1

**Comment by sneko** [comment]
@ndelangen what does it imply? This PR can now be merged with just a few changes or is there some reasons to wait a few more weeks? Thanks

**Comment by Hypnosphi** [comment]
At the very least, we should get rid of the stage presets

**Comment by pksunkara** [comment]
I think a 4.0.0-beta release is the correct way to go once this is merged.

File: .babelrc
Comment (danielduan) [inline .babelrc]: this no longer exists on RC versions

**Comment by danielduan** [comment]
I published `babel-plugin-react-docgen@2.0.0-babel7.0` for the upgrade. Please integrate it if you can and let me know if it doesn't work.

**Comment by ndelangen** [comment]
I've upgraded to latest rc, fixed all deprecations

I had to regenerate the yarn lockfile as it become 'corrupt'
which in turn meant a lot of little tweaks necessary apparently

I haven't done runtime checks yet, or tested what effect his has on browser support

File: addons/actions/package.json
Comment (Hypnosphi) [inline addons/actions/package.json]: Not actually needed here, it's in resolve.alias

File: .eslintrc.js
Comment (Hypnosphi) [inline .eslintrc.js]: Is there anything left that it still disallows?

BTW this rule is about performance so we probably shouldn't shut it up like this

File: MIGRATION.md
Comment (Hypnosphi) [inline MIGRATION.md]: now it's only babel-loader

File: addons/info/src/components/types/PrettyPropType.js
Comment (Hypnosphi) [inline addons/info/src/components/types/PrettyPropType.js]: ESLint packages update seems quite irrelevant to this PR, and it makes it harder to review =(

File: app/html/package.json
Comment (Hypnosphi) [inline app/html/package.json]: Why downgrade to less secure version?

File: app/react/src/server/wrapDefaultBabelConfig.js
Comment (Hypnosphi) [reply app/react/src/server/wrapDefaultBabelConfig.js]: Actualy we can just use [`babel-preset-react-app`](https://github.com/facebook/create-react-app/tree/next/packages/babel-preset-react-app) here. @norbert WDYT?

File: MIGRATION.md
Comment (ndelangen) [reply MIGRATION.md]: done

File: lib/cli/generators/METEOR/index.js
Comment (Hypnosphi) [inline lib/cli/generators/METEOR/index.js]: I had a warning from Babel 7 that told me to add this `module:`. Is there a chance that now it runs with Babel 6 instead?

Comment (Hypnosphi) [reply lib/cli/generators/METEOR/index.js]: For some reason, there's no `babel-loader` here: https://github.com/storybooks/storybook/blob/f8332e744b56fd4182eeba86538223f8bf43f139/lib/cli/test/snapshots/meteor/package.json

Comment (Hypnosphi) [reply lib/cli/generators/METEOR/index.js]: Looks like I just forgot to change METEOR on https://github.com/storybooks/storybook/commit/3c0e3832e169e5e874c1bc774b7b68e652c1c803

Comment (Hypnosphi) [reply lib/cli/generators/METEOR/index.js]: > But this did not come up in CircleCI

This is because you changed only circleCI config when introducing e2e

Comment (ndelangen) [reply lib/cli/generators/METEOR/index.js]: Seems to be a completely unnecessary plugin that's 3 years old, and unmaintained..
If everything is green with this gone, I'm just removing it.

Comment (Hypnosphi) [reply lib/cli/generators/METEOR/index.js]: Please also add `installBabelLoader` in meteor

**Comment by shilman** [comment]
There's a lot of people clamoring for this change, so here's the 4.0 release plan, based on community sentiment (cross-posted from the `#maintenance` channel on our [slack](https://now-examples-slackin-rrirkqohko.now.sh/))

- After the babel 7 PR is merged, we’ll shift into the release process with one more alpha and then some RC’s to get everything stable
- We might squeeze in a few more small improvements, but webpack4, babel7, and cleanup of 3.x deprecations are the big breaking changes for 4.0
- I’ll be coordinating the release posts and related documentation with everybody starting this weekend

Would love your feedback and suggestions on this!

File: addons/storyshots/storyshots-core/package.json
Comment (pksunkara) [inline addons/storyshots/storyshots-core/package.json]: I don't think we need this.

File: addons/storyshots/storyshots-puppeteer/package.json
Comment (pksunkara) [inline addons/storyshots/storyshots-puppeteer/package.json]: I don't think we need this.

File: addons/storysource/package.json
Comment (pksunkara) [inline addons/storysource/package.json]: I don't think we need this.

File: app/html/package.json
Comment (pksunkara) [inline app/html/package.json]: Do we really need babel in html app? I am not actually sure why we added it to all apps.

File: app/react/src/server/wrapDefaultBabelConfig.js
Comment (pksunkara) [reply app/react/src/server/wrapDefaultBabelConfig.js]: I agree with @Hypnosphi

File: app/vue/package.json
Comment (pksunkara) [inline app/vue/package.json]: I don't think we need this.

File: lib/cli/test/fixtures/meteor/package.json
Comment (pksunkara) [inline lib/cli/test/fixtures/meteor/package.json]: I don't think we need this.

File: lib/core/package.json
Comment (pksunkara) [reply lib/core/package.json]: Yup, **we** don't use it. So, we don't need it.

**Review by pksunkara — COMMENTED** [review]
Just a few things I think we forgot to change.

File: addons/storysource/package.json
Comment (pksunkara) [reply addons/storysource/package.json]: Didnt realize that, thanks

**Comment by Hypnosphi** [comment]
Just not to forget: https://github.com/storybooks/storybook/pull/3746#discussion_r211027591

File: .teamcity/OpenSourceProjects_Storybook/buildTypes/OpenSourceProjects_Storybook_CliTest.kt
Comment (Hypnosphi) [inline .teamcity/OpenSourceProjects_Storybook/buildTypes/OpenSourceProjects_Storybook_CliTest.kt]: Looks like `yarn test --cli` won't work without local registry anyway, so maybe it should start it always?

Comment (ndelangen) [reply .teamcity/OpenSourceProjects_Storybook/buildTypes/OpenSourceProjects_Storybook_CliTest.kt]: yeah that's a fair point.

**Review by pksunkara — APPROVED** [review]

**Comment by igor-dv** [comment]
Don't we need to add `@babel/core` to every example now?

**Comment by Hypnosphi** [comment]
> Don't we need to add @babel/core to every example now?

Technically we do (and babel-loader if it's missing)

**Comment by ndelangen** [comment]
LGTM @igor-dv

**Comment by igor-dv** [comment]
Done - https://github.com/storybooks/storybook/pull/4062

**Comment by sebinsua** [comment]
Will you publish a new alpha version containing this?

**Comment by ndelangen** [comment]
@shilman shall we release an alpha version today?

**Comment by shilman** [comment]
Released: https://github.com/storybooks/storybook/releases/tag/v4.0.0-alpha.17

**Comment by chawax** [comment]
I tried this release to make Storybook work with React Native 0.56.0 but I always have a message about unallowed presets. Is there something to change in our project ? A working example somewhere ?

The error message I have when starting Storybook :

```
ERROR in ./storybook/addons.js
Module build failed (from ./node_modules/babel-loader/lib/index.js):
Error: Plugin/Preset files are not allowed to export objects, only functions. In /Users/administrateur/xxx/xxx-app/node_modules/babel-preset-react/lib/index.js
    at createDescriptor (/Users/administrateur/xxx/xxx-app/node_modules/@babel/core/lib/config/config-descriptors.js:178:11)
    at items.map (/Users/administrateur/xxx/xxx-app/node_modules/@babel/core/lib/config/config-descriptors.js:109:50)
    at Array.map (<anonymous>)
    at createDescriptors (/Users/administrateur/xxx/xxx-app/node_modules/@babel/core/lib/config/config-descriptors.js:109:29)
    at createPresetDescriptors (/Users/administrateur/xxx/xxx-app/node_modules/@babel/core/lib/config/config-descriptors.js:101:10)
    at passPerPreset (/Users/administrateur/xxx/xxx-app/node_modules/@babel/core/lib/config/config-descriptors.js:58:96)
    at cachedFunction (/Users/administrateur/xxx/xxx-app/node_modules/@babel/core/lib/config/caching.js:33:19)
    at presets.presets (/Users/administrateur/xxx/xxx-app/node_modules/@babel/core/lib/config/config-descriptors.js:29:30)
    at mergeChainOpts (/Users/administrateur/xxx/xxx-app/node_modules/@babel/core/lib/config/config-chain.js:315:26)
    at /Users/administrateur/xxx/xxx-app/node_modules/@babel/core/lib/config/config-chain.js:278:7
 @ multi ./storybook/addons.js ./node_modules/@storybook/react-native/dist/manager/index.js manager[0]
```

**Comment by Hypnosphi** [comment]
We didn't migrate React Native part yet

**Comment by chawax** [comment]
OK, thanks for your response. Any idea when it will be migrated ?

**Comment by chrisnojima** [comment]
fyi RN 0.57 (in rc) supports babel 7

**Comment by Hypnosphi** [comment]
0.56 does as well

**Comment by Hypnosphi** [comment]
@Gongreg any ideas?

**Comment by chawax** [comment]
I just saw migrating to RN 0.56.0 is not a priority for Expo team : https://blog.expo.io/expo-sdk-v29-0-0-is-now-available-f001d77fadf

Absolutely no chance to see Storybook React Native migrating to Babel 7 before CRNA and Expo do it ?

**Pull request merged**
